[Update] Support update syntax

EmmyMiao87 opened the pull request:
## Proposed changes

    [Update] Support update syntax

    The current update syntax only supports updating the filtered data of a single table.

    Syntax:

     * UPDATE table_reference
     *     SET assignment_list
     *     [WHERE where_condition]
     *
     * value:
     *     {expr}
     *
     * assignment:
     *     col_name = value
     *
     * assignment_list:
     *     assignment [, assignment] ...

    Example
    Update unique_table
         set v1=1
         where k1=1

    New Frontend Config: enable_concurrent_update
    This configuration is used to control whether multi update stmt can be executed concurrently in one table.
    Default value is false which means A table can only have one update task being executed at the same time.
    If users want to update the same table concurrently,
      they need to modify the configuration value to true and restart the master frontend.
    Concurrent updates may cause write conflicts, the result is uncertain, please be careful.

    The main realization principle:
    1. Read the rows that meet the conditions according to the conditions set by where clause.
    2. Modify the result of the row according to the set clause.
    3. Write the modified row back to the table.

    Some restrictions on the use of update syntax.
    1. Only the unique table can be updated
    2. Only the value column of the unique table can be updated
    3. The where clause currently only supports single tables

    Possible risks:
    1. Since the current implementation method is a row update,
         when the same table is updated concurrently, there may be concurrency conflicts which may cause the incorrect result.
    2. Once the conditions of the where clause are unsatisfactory, it is likely to cause a full table scan and affect query performance.
       Please pay attention to whether the column in the where clause can match the index when using it.

    [Docs][Update] Add update document and sql-reference

    Fixed #6229

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)
- [x] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #6229) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If these changes need document changes, I have updated the document
- [x] Any dependent changes have been merged

EmmyMiao87 committed:
Fix ut error

morningman commented on fe/fe-core/src/main/java/org/apache/doris/planner/StreamLoadPlanner.java:
```diff
@@ -203,7 +203,7 @@ public TExecPlanFragmentParams plan(TUniqueId loadId) throws UserException {
     // get all specified partition ids.
-    // if no partition specified, return all partitions
+    // if no partition specified, return null
     private List<Long> getAllPartitionIds() throws DdlException, AnalysisException {
```
You change it to return null, but don' t change the caller to handle null? Such as in `StreamLoadPlanner`?

EmmyMiao87 replied:
I unified handle this in the OlapTableSink.

morningman commented on docs/en/sql-reference/sql-statements/Data Manipulation/UPDATE.md:
```diff
@@ -0,0 +1,75 @@
+---
```
This document also need to be added to the sidebar

EmmyMiao87 replied:
done

morningman commented on docs/zh-CN/administrator-guide/update.md:
```diff
@@ -0,0 +1,125 @@
+如果我们需要修改或更新 Doris 中的数据，就可以使用 UPDATE 命令来操作。
+
+## 适用场景
+
```
使用场景可以写明需要 unique key

EmmyMiao87 replied:
done

morningman commented on docs/zh-CN/administrator-guide/update.md:
```diff
@@ -0,0 +1,125 @@
+	| 1 | 100| 待付款 |
+        | 1 | 100 | 待发货 |
+	由于表 order 是 UNIQUE 模型，所以相同 Key 的行，之后后者才会生效，所以最终效果如下：
+	|订单id | 订单金额| 订单状态|
```
use space instead of tab

EmmyMiao87 replied:
done

EmmyMiao87 commented on docs/zh-CN/administrator-guide/update.md:
```diff
@@ -0,0 +1,125 @@
+
+## 版本
+
+Doris Version 0.14.7 +
```
done

EmmyMiao87 committed:
Fix doc

EmmyMiao87 committed:
[Update] Support update syntax

The current update syntax only supports updating the filtered data of a single table.

Syntax:

 * UPDATE table_reference
 *     SET assignment_list
 *     [WHERE where_condition]
 *
 * value:
 *     {expr}
 *
 * assignment:
 *     col_name = value
 *
 * assignment_list:
 *     assignment [, assignment] ...

Example
Update unique_table
     set v1=1
     where k1=1

New Frontend Config: enable_concurrent_update
This configuration is used to control whether multi update stmt can be executed concurrently in one table.
Default value is false which means A table can only have one update task being executed at the same time.
If users want to update the same table concurrently,
  they need to modify the configuration value to true and restart the master frontend.
Concurrent updates may cause write conflicts, the result is uncertain, please be careful.

The main realization principle:
1. Read the rows that meet the conditions according to the conditions set by where clause.
2. Modify the result of the row according to the set clause.
3. Write the modified row back to the table.

Some restrictions on the use of update syntax.
1. Only the unique table can be updated
2. Only the value column of the unique table can be updated
3. The where clause currently only supports single tables

Possible risks:
1. Since the current implementation method is a row update,
     when the same table is updated concurrently, there may be concurrency conflicts which may cause the incorrect result.
2. Once the conditions of the where clause are unsatisfactory, it is likely to cause a full table scan and affect query performance.
   Please pay attention to whether the column in the where clause can match the index when using it.

[Docs][Update] Add update document and sql-reference

Fixed #6229

morningman commented on fe/fe-core/src/main/cup/sql_parser.cup:
```diff
@@ -3114,6 +3117,14 @@ insert_source ::=
 
+// update stmt
+update_stmt ::=
+    KW_UPDATE table_name:tbl KW_SET expr_list:setExprs where_clause:whereClause
```
You forgot to add `KW_UPDATE`

EmmyMiao87 committed:
Fix error

EmmyMiao87 committed:
Fix ut

morningman reviewed (dismissed):
LGTM
